### ISSUE TITLE
flex-direction column with incomplete flex shorthand works differently in IE than Chrome

### ISSUE THREAD

**Issue by joefiorini** [philipwalton/flexbugs#58]
I ran into an issue in IE where any flex children inside a flex-direction column container would lose their heights in IE11. I started out using the explicit height fix, until I ran into a child that needed to flex based on the height of its children.

I tried a couple of solutions from here, nothing worked. Then when I looked at my CSS I noticed that I was using `flex: 1;` anywhere I used the flex shorthand. I went through and changed them all to `flex: 0 auto` and that fixed it.

Why?

After a _little_ digging (haven't done any spec reading yet) I discovered that when using flex shorthand without an explicit value for `flex-basis`, Chrome defaults it to `0%`, but IE11 defaults to `0px`. Therefore we need to make sure IE keeps the `auto` value for `flex-basis`. (One weird thing here: in Chrome `flex-basis: 0%` works the same as `flex-basis: auto` but in IE11 `flex-basis: 0%` works the same as `flex-basis: 0px`)

It seems like a good idea to _always_ use `1 0 auto` as a default for `flex` shorthand, which I'm sure is a better practice than what I was doing anyway.

**Comment by jpdevries** [comment]
> It seems like a good idea to always use 1 0 auto as a default for flex shorthand, which I'm sure is a better practice than what I was doing anyway.

Technically, I don't think you've done anything wrong. The spec repeatedly advises using the shorthand syntax but as you point out the browser vendor aren't all in agreement with the defaults so the `auto` is needed.

> Authors are encouraged to control flexibility using the flex shorthand rather than with its longhand properties directly, as the shorthand correctly resets any unspecified components to accommodate common uses. - [7.2. Components of Flexibility](http://www.w3.org/TR/css-flexbox-1/#flex-components)

##### Also see
 - [7.1.1 Common Values of 'flex'](http://www.w3.org/TR/css-flexbox-1/#flex-common)
 - unitless flex-basis in [Philip's post](http://philipwalton.com/articles/normalizing-cross-browser-flexbox-bugs/).

**Comment by philipwalton** [comment]
> One weird thing here: in Chrome `flex-basis: 0%` works the same as `flex-basis: auto`

That's not true. They render the same in your example, but they are actually very different. This [diagram](http://www.w3.org/TR/css-flexbox-1/#valdef-flex-flex-basis) shows what the difference is.

I don't think this bug has anything to do with the differing `flex-basis` values. I think it's actually a complex blend of flexbug 1 (not honoring the minimum content size) mixed with IE11 incorrectly calculating height (though it's hard to say for sure). If it were just flexbug 1, then using `flex-shrink: 0` would fix it, but as you can see, it doesn't.

This also only seems to affect the vertical direction, and in practice I don't think it's very common to use `flex-basis` values other than `auto` inside non-fixed-height containers.

**Comment by jpdevries** [comment]
> This also only seems to affect the vertical direction, and in practice I don't think it's very common to use flex-basis values other than auto inside non-fixed-height containers.

Flexbox in the wild still isn't very common but I think as it becomes more used for entire page layouts people will run into these issues more often.

For example, this is a pretty common layout for web apps
![](http://j4p.us/image/2k1U2k370o12/Screen%20Shot%202015-06-12%20at%2010.52.44%20AM.png)

According to the spec, you should be able to set the same `flex-basis` on the "header" of each of those columns to keep them the same height.

```scss
.col header {
  flex:0 0 4em; //dont grow dont shrink try and be 4em tall
}
```

_Note this requires a flex-basis other than auto inside a non-fixed height container_

If the don't come out the same height, and those borders don't line up correctly an Art Director might flip your desk over.

**Comment by philipwalton** [comment]
Well, maybe the word "fixed-height" is ambiguous. I probably should have said "definite height".

By definite height, I mean the height (or max/min height) properties are either set or determinate (e.g. by another flex container). Your example seems to be a height of 100%, so I'd consider that definite.

**Comment by jpdevries** [comment]
> Your example seems to be a height of 100%, so I'd consider that definite.

if there was a footer within that component though would it still be definite height? Say you want the footer to be a certain height, and the flex parent which holds the two columns to fill the remaining space.

**Comment by philipwalton** [comment]
> and the flex parent which holds the two columns to fill the remaining space.

If the parent is filling the remaining space, then the parent would be `height: 100%` or `min-height: 100%` or something like that.

Keep in mind that in order for the flex grow/shrink properties to have any meaning, there needs to be a definite width or height that they're growing or shrinking to. In the column direction, if there's no definite height, then there's really no reason to use `display: flex` over `display: block`.

**Comment by hanhpho** [comment]
This also happens in Safari 8, OS X Yosemite.
